File: fsharp/rail-fence-cipher/RailFenceCipher.fs
module RailFenceCipher
open System

let private index rails i = 
    let adj = rails - 1
    let index = i % (adj * 2)
    if index > adj then adj - (index - adj) else index      

let private segmentString lengths (input : string) =
    let segments = Seq.zip (Seq.scan (+) 0 lengths) lengths
    Seq.map (input.Substring : int * int -> string) segments

let encode rails (message: string) = 
    message 
        |> Seq.mapi (fun i c -> (index rails i, c))
        |> Seq.groupBy (fun (i,_) -> i) 
        |> Seq.map (fun (_,cs) -> cs |> Seq.map (fun (_,c) -> c) |> String.Concat)
        |> String.Concat

let decode rails (message: string) = 
    let rows = 
        [0..message.Length - 1] 
        |> Seq.map (fun i -> (i, index rails i))
        |> Seq.groupBy (fun (_,l) -> l)
    let segments = message |> segmentString (rows |> Seq.map (fun (_,s) -> Seq.length s))
    let chars = Seq.map2 (fun (_,cs) line -> Seq.map2 (fun (i,_) c -> (i,c)) cs line) rows segments
    chars |> Seq.concat |> Seq.sortBy (fun (i,_) -> i) |> Seq.map snd |> String.Concat

File: go/rail-fence-cipher/rail_fence_cipher.go
package railfence

import "strings"

func Encode(text string, rails int) string {
	text = strings.ReplaceAll(text, " ", "")
	lines := make([]string, rails)
	r, d := 0, 1
	for _, c := range text {
		lines[r] += string(c)
		if d == 1 && r == len(lines)-1 {
			d = -1
		} else if d == -1 && r == 0 {
			d = 1
		}
		r += d
	}
	result := ""
	for _, line := range lines {
		result += line
	}
	return result
}

func Decode(cipher string, rails int) string {
	lines := make([][]int, rails)
	r, d := 0, 1
	for i := 0; i < len(cipher); i++ {
		lines[r] = append(lines[r], i)
		if d == 1 && r == len(lines)-1 {
			d = -1
		} else if d == -1 && r == 0 {
			d = 1
		}
		r += d
	}
	result := make([]byte, len(cipher))
	i := 0
	for _, line := range lines {
		for _, index := range line {
			result[index] = cipher[i]
			i++
		}
	}
	return string(result)
}
